content/blog/2019-11-13-gep.md/index.md

```diff
@@ -0,0 +1,458 @@
+
+As mentioned previously, the former is difficult to deal with
+as an IR compiler pass since a general solution would require
+modifying function signatures to pass allocation size information.
+
+The latter problem stems primarily from three operations: `load`, `store` and `getelementptr`.
+While complications arising from GEPs are straightforward to solve (since that's already an instruction
+we're instrumenting already), without precise memory dependency and alias analyses,
+we cannot track allocation sizes of pointers derived from other pointers.
+The most common case we noticed, anecdotally,
+were global pointers to data that were "malloced" at the beginning of the `main` function,
+but accessed throughout the program.
+
+For instance, take the **Blackscholes** benchmark, for which we instrumented *no* GEP instructions.
+It has a global pointer to an array of `floats` called `prices`, whose size is determined during
+the beginning of execution:
+
+```C
+fptype *prices;
+...
+main() {
+   ...
+   prices = (fptype*)malloc(numOptions*sizeof(fptype));
+   ...
+}
+```
+
+The above allocation translates to the following LLVM IR:
+```C
+%46 = tail call noalias i8* @malloc(i64 %45) #9
+store i8* %46, i8** bitcast (float** @prices to i8**), align 8
+```
+
+Our analysis determined that the runtime size of the memory pointed to
+by `%46` was given by the value `%45`.
+However, `%46` is not used as the argument to any GEP instruction, instead
+later operations use a `load` to retrieve the array pointer from `prices`.
+
+```C
+%179 = load float*, float** @prices, align 8
+%180 = getelementptr inbounds float, float* %179, i64 %159
+```
+
+Since we are not running a memory dependency analysis we could
+not determine that the size of the allocation pointed to by `%179` was `%45`.
+
+Often, these two problems combined, since `prices` may be accessed
+outside the scope at which `%45` is available and we therefore would need
+to modify the program to communicate this allocation information (potentially
+via an extra global variable).
+
+
+### Evaluation: Overhead
+
+We measured runtime overhead in terms of wall clock time purely because it
+was the simplest thing to instrument. In the following graph we report the
+average slowdown caused by our instrumentation (lower is better). At the end
+of this section is a graph reporting our base results (rather than the ratio)
+which reports the mean runtime for both configurations. Error bars on that
+graph represent one standard deviation.
+
+<img src="overhead.png"/>
+
+In the case of the Ferret benchmark, our instrumentation caused the implementation
+to exit prematurely. Since the benchmark is quite large we did not have time to investigate
+why this was; it is possible that Ferret intentionally executes "unsafe" GEP instructions.
+Since our instrumentation did not cause bugs in any of the other implementations we find
+this to be a likely cause, but it does warrant further examination.
```

Comment:
It would indeed be interesting to investigate… something like Address Sanitizer could help pin down the line of code where a safety violation *might* happen. It's unfortunately not unheard of for benchmarks to be full of UB.